proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+   synonyms may not be parameterized.
+
+#. The form ``decl → 'type' tyvar '=' type`` is not allowed at top-level.
+
+#. These new declaration forms introduce local type synonyms in terms, which scope over the same
+   region of code that other declarations in the same ``let`` / ``where`` clause scope over.
+
+   Like other type synonyms, local type synonyms may not be recursive.
+
+#. Wildcards are allowed in the right-hand side of local synonyms. At usage sites of the
+   synonym, the synonym is expanded. It is an error if that location does not allow wildcards.
+   The wildcard is understood to stand for just one type shared among all the expansions.
+
+Effects
+^^^^^^^
+
+1. We can now bind local type synonyms, avoiding the need to do so via pattern or result
+   signatures.
+
+#. One challenge is how to present these local synonyms in error messages. It might be
+   best to aggressively expand (unlike top-level type synonyms), especially because these
+   local synonyms might refer to other local type variables that are in scope. As we gain
+   experience with this new form, we can refine their appearance in error messages.
+
+#. Note that this proposal does *not* allow for top-level lower-case type synonyms. There
+   is nothing stopping us from doing so, but it would seem to violate expectations of Haskellers
+   and would be the first instance of a lower-case type variable being in scope at the top level.
+
+``let`` in patterns
+~~~~~~~~~~~~~~~~~~~
+
+This part of this proposal allows introducing a ``let``\ -binding in a pattern.
+The bound variable(s) scope over the same region of code as the pattern-bound
+variables do.
+
+The syntax for this feature is a bit awkward, and so this proposal presents
+two alternatives for discussion.
+
+Motivation
+^^^^^^^^^^
+
+1. A careful reader will note that allowing ``let`` for `type synonyms <#type-let>`_
+   does not, by itself, replace a binding such as that in ``f (True :: b) = ...``
+   because the current form binds the type variable in the pattern. This part
+   of the proposal closes this gap. See `examples <#let-in-pattern-example>`_ below.
+
+#. Though admittedly a weakish motivation, there is currently no way to share
+   expressions used in common in multiple view patterns. See `examples <#let-in-pattern-example>`_
+   below.
+
+Specification Alternative 1
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. With ``-XExtendedLet``, add a new form of pattern as follows::
+
+     pat → 'let' decls 'in' pat
+
+#. Any entites bound in ``decls`` scope over the same region of the program
+   that pattern-bound variables scope over, with the addition of the ``decls``
+   themselves (that is, the declarations can be recursive).
+
+Specification Alternative 2
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+0. **Background**. Here are some productions from the `Haskell 2010 Report`_::
+
+     funlhs → var apat {apat}
+            | pat varop pat
+            | '(' funlhs ')' apat {apat}
```

Comment:
Oh wow… TIL.

Though anybody knows why it exists?

Reply:
Maybe `(f a b) c d` is the same as `f a b c d` in expressions, and so they should be the same in a declaration LHS? Or maybe it's to allow the otherwise-sensible

```hs
(f . g) x = f (g x)
```